examples/diagram.dx

```diff
@@ -0,0 +1,193 @@
+  totalSize = sum for i. listLength lists.i
+  AsList _ $ fst $ withState 0 \listIdx.
+    fst $ withState 0 \eltIdx.
+      for i:(Fin totalSize).
+        while (\(). get eltIdx >= listLength (lists.((get listIdx)@_))) \().
+          eltIdx := 0
+          listIdx := get listIdx + 1
+        (AsList _ xs) = lists.((get listIdx)@_)
+        eltIdxVal = get eltIdx
+        eltIdx := eltIdxVal + 1
+        xs.(eltIdxVal@_)
+
+
+Point : Type = (Float & Float)
+
+data Geom =
+  Spot
+  Circle Float
+  Rectangle Float Float  -- width, height
+  Line Point
+
+-- HTML color string. TODO: consider an RGB triple instead
+Color : Type = String
+
+-- TODO: we should add overloaded string literals so we don't need this
+def str (n:Int) ?-> (s:(Fin n=>Char)) : String = AsList _ s
+
+black : Color = str "black"
+white : Color = str "white"
+gray  : Color = str "gray"
+red   : Color = str "red"
+green : Color = str "green"
+blue  : Color = str "blue"
+
+Attributes : Type = { fill   : Maybe Color
+                    & stroke : Maybe Color
+                    & lw     : Int }
+defaultAttributes : Attributes = { fill   = Nothing
+                                 , stroke = Just black
+                                 , lw = 1 }
+
+-- TODO: consider sharing attributes among a set of objects for efficiency
+data Diagram = MkDiagram (List (Attributes & Point & Geom))
+
+instance monoidDiagram : Monoid Diagram where
+  mempty = MkDiagram mempty
+  mcombine = \(MkDiagram d1) (MkDiagram d2). MkDiagram $ d1 <> d2
+
+def concatDiagrams (diagrams:n=>Diagram) : Diagram =
+  MkDiagram $ concat for i.
+    (MkDiagram d) = diagrams.i
+    d
+
+-- TODO: arbitrary affine transformations. Our current representation of
+-- rectangles and circles means we can only do scale/flip/rotate90.
+-- Should we use lenses/isomorphisms for these instead?
+def applyTransformation
+      (transformPoint: Point -> Point)
+      (transformGeom:  Geom -> Geom)
+      (d:Diagram) : Diagram =
+  (MkDiagram (AsList _ objs)) = d
+  (MkDiagram $ AsList _ for i.
+    (attr, p, geom) = objs.i
+    (attr, transformPoint p, transformGeom geom))
+
+flipY : Diagram -> Diagram =
+  applyTransformation (\(x,y). (x, -y)) \geom. case geom of
+    Spot          -> Spot
+    Circle r      -> Circle r
+    Rectangle w h -> Rectangle w h
+    Line (x, y)   -> Line (x, -y)
+
+def scale (s:Float) : (Diagram -> Diagram) =
+  applyTransformation ( \(x,y). (s * x, s * y) ) \geom. case geom of
+    Spot          -> Spot
+    Circle r      -> Circle (s * r)
+    Rectangle w h -> Rectangle (s * w) (s * h)
+    Line (x, y)   -> Line (s * x, s * y)
+
+def moveXY ((offX, offY) : Point) : (Diagram -> Diagram) =
+  applyTransformation (\(x,y). (x + offX, y + offY) ) id
+
+def singletonDefault (geom:Geom) : Diagram =
+  MkDiagram $ AsList _ [(defaultAttributes, (0.0, 0.0), geom)]
+
+-- TODO: should really allow nullary functions with `def`
+spot                           : Diagram = singletonDefault Spot
```

Comment:
nit: I saw this function being used in a different file and was very confused as to what `spot` means. Maybe there's a clearer name. Although I guess it's nice to have a short name. Maybe it's just that `spot` isn't obviously a geometric primitive (for me at least)?

Reply:
I like the idea of adding some disambiguation word to it. Maybe one of:

- `spotDiagram` (and `SpotGeom`)
- `dotDiagram` or `drawDot` (and `DotGeom`)
- `drawPoint` (and `PointGeom`)